File: app/src/main/java/com/kinwatt/powermeter/data/provider/SensorProvider.java
package com.kinwatt.powermeter.data.provider;

import android.content.Context;
import android.hardware.Sensor;
import android.widget.Toast;

import com.kinwatt.powermeter.data.SensorData;
import com.kinwatt.powermeter.data.ServiceData;
import com.kinwatt.powermeter.data.mappers.SensorMapper;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class SensorProvider {

    private static SensorProvider instance;

    private Context context;

    private File dataFile;

    private List<SensorData> sensors = new ArrayList<>();

    private SensorProvider(Context context) {
        this.context = context;

        dataFile = new File(context.getFilesDir(), "devices.json");
        if (dataFile.exists()) {
            try {
                for (SensorData data : SensorMapper.load(dataFile)) {
                    if (!sensors.contains(data)) {
                        sensors.add(data);
                    }
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    public List<SensorData> getAll() {
        return sensors;
    }

    public SensorData findSensor(UUID serviceId) {
        for (SensorData sensor : sensors) {
            for (ServiceData service : sensor.getServices()) {
                if (service.getUuid().equals(serviceId)) {
                    return sensor;
                }
            }
        }
        return null;
    }

    public void add(SensorData sensor) {
        if (!this.sensors.contains(sensor)) {
            this.sensors.add(sensor);
            try {
                SensorMapper.save(sensors, dataFile);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    public boolean remove(SensorData sensor) {
        boolean res = this.sensors.remove(sensor);
        try {
            SensorMapper.save(sensors, dataFile);
        } catch (IOException e) {
            e.printStackTrace();
            res = false;
        }
        return res;
    }

    public static SensorProvider getProvider(Context context) {
        if (instance == null) {
            instance = new SensorProvider(context);
        }
        return instance;
    }
}

File: app/src/main/java/com/kinwatt/powermeter/data/provider/SensorProvider.kt
package com.kinwatt.powermeter.data.provider

import android.content.Context
import android.hardware.Sensor
import android.widget.Toast

import com.kinwatt.powermeter.data.SensorData
import com.kinwatt.powermeter.data.ServiceData
import com.kinwatt.powermeter.data.mappers.SensorMapper

import java.io.File
import java.io.IOException
import java.util.ArrayList
import java.util.UUID

class SensorProvider private constructor(private val context: Context) {

    private val dataFile: File

    private val sensors = ArrayList<SensorData>()

    val all: List<SensorData>
        get() = sensors

    init {

        dataFile = File(context.filesDir, "devices.json")
        if (dataFile.exists()) {
            try {
                for (data in SensorMapper.load(dataFile)) {
                    if (!sensors.contains(data)) {
                        sensors.add(data)
                    }
                }
            } catch (e: IOException) {
                e.printStackTrace()
            }

        }
    }

    fun findSensor(serviceId: UUID): SensorData? =
        sensors.flatMap { sensor -> sensor.services.map { Pair(sensor, it) } }
               .firstOrNull{ it.second.uuid == serviceId }?.first

    fun add(sensor: SensorData) {
        if (!this.sensors.contains(sensor)) {
            this.sensors.add(sensor)
            try {
                SensorMapper.save(sensors, dataFile)
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    fun remove(sensor: SensorData): Boolean {
        var res = this.sensors.remove(sensor)
        try {
            SensorMapper.save(sensors, dataFile)
        } catch (e: IOException) {
            e.printStackTrace()
            res = false
        }

        return res
    }

    companion object {

        private var instance: SensorProvider? = null

        fun getProvider(context: Context): SensorProvider {
            if (instance == null) {
                instance = SensorProvider(context)
            }
            return instance!!
        }
    }
}
